client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -108,7 +108,30 @@ public void readerOffline(String readerId, Position lastPosition) {
 
     @Override
     public CompletableFuture<Checkpoint> initiateCheckpoint(String checkpointName, ScheduledExecutorService backgroundExecutor) {
-        synchronizer.updateStateUnconditionally(new CreateCheckpoint(checkpointName));
+
+        String rejectMessage = "rejecting checkpoint request since pending checkpoint reaches max allowed limit";
+
+        boolean canPerformCheckpoint = synchronizer.updateState((state, updates) -> {
+            ReaderGroupConfig config = state.getConfig();
+            CheckpointState checkpointState = state.getCheckpointState();
+            int maxOutstandingCheckpointRequest = config.getMaxOutstandingCheckpointRequest();
+            int currentOutstandingCheckpointRequest = checkpointState.getOutstandingCheckpoints();
+            if (currentOutstandingCheckpointRequest >= maxOutstandingCheckpointRequest) {
+                String errorMessage = rejectMessage + maxOutstandingCheckpointRequest;
```

Comment:
Rather than doing string concat here, you can pass it to the templated warn method below which will do it all at once.